File: Etapa3-Construcao/3.1-Stack1_Node_TS/src/domain/entities/Quarto.ts
import { TipoQuarto } from '../enums/TipoQuarto.js';
import { Disponibilidade } from '../enums/Disponibilidade.js';
import { NumeroQuarto } from '../value-objects/NumeroQuarto.js';
import { PrecoHora } from '../value-objects/PrecoHora.js';
import { CamaQuarto } from './CamaQuarto.js';
import { TipoCama } from '../enums/TipoCama.js';

// Entidade Quarto: encapsula invariantes e comportamentos do quarto.
// Comentários explicam decisões SOLID/DDD.
export class Quarto {
  // Mantemos id opcional por ora (repositório atribui se necessário)
  private _id?: string;
  private _numero: NumeroQuarto;
  private _capacidade: number;
  private _tipo: TipoQuarto;
  private _precoHora: PrecoHora;
  private _frigobar: boolean;
  private _cafeManha: boolean;
  private _arCondicionado: boolean;
  private _tv: boolean;
  private _disponibilidade: Disponibilidade;
  private _camas: CamaQuarto[] = [];
  private _ativo = true; // soft delete

  private constructor(params: {
    id?: string;
    numero: NumeroQuarto;
    capacidade: number;
    tipo: TipoQuarto;
    precoHora: PrecoHora;
    frigobar: boolean;
    cafeManha: boolean;
    arCondicionado: boolean;
    tv: boolean;
    disponibilidade?: Disponibilidade;
    camas?: CamaQuarto[];
  }) {
    this._id = params.id;
    this._numero = params.numero;
    this._capacidade = Quarto.validarCapacidade(params.capacidade);
    this._tipo = params.tipo;
    this._precoHora = params.precoHora;
    this._frigobar = params.frigobar;
    this._cafeManha = params.cafeManha;
    this._arCondicionado = params.arCondicionado;
    this._tv = params.tv;
    this._disponibilidade = params.disponibilidade ?? Disponibilidade.LIVRE;
    this._camas = params.camas ? [...params.camas] : [];
  }

  // Fábrica estática centraliza regras de criação (SRP) e mantém invariantes desde o início.
  static criar(params: {
    id?: string;
    numero: NumeroQuarto;
    capacidade: number;
    tipo: TipoQuarto;
    precoHora: PrecoHora;
    frigobar: boolean;
    cafeManha: boolean;
    arCondicionado: boolean;
    tv: boolean;
    disponibilidade?: Disponibilidade;
    camas?: CamaQuarto[];
  }): Quarto {
    return new Quarto(params);
  }

  // Métodos de negócio explícitos promovem encapsulamento (EVITA set público indiscriminado)
  atualizarDados(parciais: Partial<{
    capacidade: number;
    tipo: TipoQuarto;
    precoHora: PrecoHora;
    frigobar: boolean;
    cafeManha: boolean;
    arCondicionado: boolean;
    tv: boolean;
  }>): void {
    if (parciais.capacidade !== undefined) this._capacidade = Quarto.validarCapacidade(parciais.capacidade);
    if (parciais.tipo !== undefined) this._tipo = parciais.tipo;
    if (parciais.precoHora !== undefined) this._precoHora = parciais.precoHora;
    if (parciais.frigobar !== undefined) this._frigobar = parciais.frigobar;
    if (parciais.cafeManha !== undefined) this._cafeManha = parciais.cafeManha;
    if (parciais.arCondicionado !== undefined) this._arCondicionado = parciais.arCondicionado;
    if (parciais.tv !== undefined) this._tv = parciais.tv;
  }

  alterarDisponibilidade(nova: Disponibilidade): void {
    this._disponibilidade = nova;
  }

  definirCamas(tipos: TipoCama[]): void {
    // substitui o conjunto de camas (operação idempotente)
    this._camas = tipos.map(t => new CamaQuarto(t));
  }

  desativar(): void { this._ativo = false; }

  private static validarCapacidade(cap: number): number {
    if (!Number.isInteger(cap) || cap <= 0 || cap > 20) throw new Error('Capacidade inválida');
    return cap;
  }

  // Getters somente leitura garantem imutabilidade externa
  get id(): string | undefined { return this._id; }
  get numero(): NumeroQuarto { return this._numero; }
  get capacidade(): number { return this._capacidade; }
  get tipo(): TipoQuarto { return this._tipo; }
  get precoHora(): PrecoHora { return this._precoHora; }
  get frigobar(): boolean { return this._frigobar; }
  get cafeManha(): boolean { return this._cafeManha; }
  get arCondicionado(): boolean { return this._arCondicionado; }
  get tv(): boolean { return this._tv; }
  get disponibilidade(): Disponibilidade { return this._disponibilidade; }
  get camas(): ReadonlyArray<CamaQuarto> { return this._camas; }
  get ativo(): boolean { return this._ativo; }
}

File: Etapa3-Construcao/3.2-Stack2_Java_Spring/src/main/java/com/hotel/quartos/domain/Quarto.java
package com.hotel.quartos.domain;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

// Entidade Quarto com invariantes básicas e métodos de negócio.
// Comentários explicam decisões (DDD light + SOLID).
@Entity
@Table(name = "quarto", uniqueConstraints = @UniqueConstraint(columnNames = "numero"))
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Quarto {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id;

    @Column(nullable = false)
    private Integer numero; // VO simplificado para exemplo; unicidade garantida por constraint

    @Column(nullable = false)
    private Integer capacidade;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private TipoQuarto tipo;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal precoHora;

    @Column(nullable = false)
    private boolean frigobar;
    @Column(nullable = false)
    private boolean cafeManha;
    @Column(nullable = false)
    private boolean arCondicionado;
    @Column(nullable = false)
    private boolean tv;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private Disponibilidade disponibilidade = Disponibilidade.LIVRE;

    @OneToMany(cascade = CascadeType.ALL, orphanRemoval = true)
    @JoinColumn(name = "quarto_id") // FK mantida em CamaQuarto
    private List<CamaQuarto> camas = new ArrayList<>();

    @Column(nullable = false)
    private boolean ativo = true; // soft delete

    @Builder
    private Quarto(Integer numero,
                  Integer capacidade,
                  TipoQuarto tipo,
                  BigDecimal precoHora,
                  boolean frigobar,
                  boolean cafeManha,
                  boolean arCondicionado,
                  boolean tv,
                  List<CamaQuarto> camas) {
        validarNumero(numero);
        this.numero = numero;
        this.capacidade = validarCapacidade(capacidade);
        this.tipo = tipo;
        this.precoHora = validarPreco(precoHora);
        this.frigobar = frigobar;
        this.cafeManha = cafeManha;
        this.arCondicionado = arCondicionado;
        this.tv = tv;
        if (camas != null) this.camas.addAll(camas);
    }

    // Regras de negócio explícitas (encapsulamento)
    public void atualizarDados(Integer capacidade, TipoQuarto tipo, BigDecimal precoHora,
                               Boolean frigobar, Boolean cafeManha, Boolean arCondicionado, Boolean tv) {
        if (capacidade != null) this.capacidade = validarCapacidade(capacidade);
        if (tipo != null) this.tipo = tipo;
        if (precoHora != null) this.precoHora = validarPreco(precoHora);
        if (frigobar != null) this.frigobar = frigobar;
        if (cafeManha != null) this.cafeManha = cafeManha;
        if (arCondicionado != null) this.arCondicionado = arCondicionado;
        if (tv != null) this.tv = tv;
    }

    public void alterarDisponibilidade(Disponibilidade nova) {
        this.disponibilidade = nova;
    }

    public void definirCamas(List<TipoCama> tipos) {
        this.camas.clear();
        tipos.forEach(t -> this.camas.add(CamaQuarto.builder().tipoCama(t).build()));
    }

    public void desativar() { this.ativo = false; }

    private static void validarNumero(Integer n) {
        if (n == null || n <= 0 || n > 9999) throw new IllegalArgumentException("Número de quarto inválido");
    }

    private static Integer validarCapacidade(Integer c) {
        if (c == null || c <= 0 || c > 20) throw new IllegalArgumentException("Capacidade inválida");
        return c;
    }

    private static BigDecimal validarPreco(BigDecimal p) {
        if (p == null || p.compareTo(BigDecimal.ZERO) <= 0) throw new IllegalArgumentException("Preço por hora inválido");
        return p.setScale(2, BigDecimal.ROUND_HALF_UP);
    }
}
